src/NHibernate/Multi/QueryBatchItemBase.cs

```diff
@@ -0,0 +1,231 @@
+
+		public virtual void Init(ISessionImplementor session)
+		{
+			Session = session;
+
+			_queryInfos = GetQueryLoadInfo();
+
+			var count = _queryInfos.Count;
+			_subselectResultKeys = new List<EntityKey[]>[count];
+			_loaderResults = new IList[count];
+
+			_finalResults = null;
+		}
+
+		/// <summary>
+		/// Gets the commands to execute for getting the not-already cached results of this query. Does retrieves
+		/// already cached results by side-effect.
+		/// </summary>
+		/// <returns>The commands for obtaining the results not already cached.</returns>
+		public IEnumerable<ISqlCommand> GetCommands()
+		{
+			for (var index = 0; index < _queryInfos.Count; index++)
+			{
+				var qi = _queryInfos[index];
+
+				if (qi.Loader.IsCacheable(qi.Parameters))
+				{
+					// Check if the results are available in the cache
+					qi.Cache = Session.Factory.GetQueryCache(qi.Parameters.CacheRegion);
+					qi.CacheKey = qi.Loader.GenerateQueryKey(Session, qi.Parameters);
+					var resultsFromCache = qi.Loader.GetResultFromQueryCache(Session, qi.Parameters, qi.QuerySpaces, qi.Cache, qi.CacheKey);
+
+					if (resultsFromCache != null)
+					{
+						// Cached results available, skip the command for them and stores them.
+						qi.Cache = null;
+						_loaderResults[index] = resultsFromCache;
+						continue;
+					}
+				}
+
+				yield return qi.Loader.CreateSqlCommand(qi.Parameters, Session);
+			}
+		}
+
+		public IEnumerable<Func<DbDataReader, int>> GetResultSetHandler()
+		{
+			var dialect = Session.Factory.Dialect;
+			List<object>[] hydratedObjects = new List<object>[_queryInfos.Count];
+
+			for (var i = 0; i < _queryInfos.Count; i++)
+			{
+				Loader.Loader loader = _queryInfos[i].Loader;
+				var queryParameters = _queryInfos[i].Parameters;
+
+				//Skip processing for items already loaded from cache
+				if (_queryInfos[i].CacheKey?.ResultTransformer != null && _loaderResults[i] != null)
+				{
+					_loaderResults[i] = loader.TransformCacheableResults(queryParameters, _queryInfos[i].CacheKey.ResultTransformer, _loaderResults[i]);
+					continue;
+				}
+
+				int entitySpan = loader.EntityPersisters.Length;
+				hydratedObjects[i] = entitySpan == 0 ? null : new List<object>(entitySpan);
+				EntityKey[] keys = new EntityKey[entitySpan];
+
+				RowSelection selection = queryParameters.RowSelection;
+				bool createSubselects = loader.IsSubselectLoadingEnabled;
+
+				_subselectResultKeys[i] = createSubselects ? new List<EntityKey[]>() : null;
+				int maxRows = Loader.Loader.HasMaxRows(selection) ? selection.MaxRows : int.MaxValue;
+				bool advanceSelection = !dialect.SupportsLimitOffset || !loader.UseLimit(selection, dialect);
+
+				var index = i;
+				yield return reader =>
+				{
+					if (advanceSelection)
+					{
+						Loader.Loader.Advance(reader, selection);
+					}
+					if (queryParameters.HasAutoDiscoverScalarTypes)
+					{
+						loader.AutoDiscoverTypes(reader, queryParameters, null);
```

Comment:
I was forgetting this one spotted in #1783, which should also be included in a PR for fixing other trouble highlighted by my previous review.